core/src/impl/Kokkos_SharedAlloc_timpl.hpp

```diff
@@ -0,0 +1,286 @@
+}
+
+template <class MemorySpace>
+void HostInaccessibleSharedAllocationRecordCommon<MemorySpace>::print_records(
+    std::ostream& s, const MemorySpace&, bool detail) {
+  (void)s;
+  (void)detail;
+#ifdef KOKKOS_ENABLE_DEBUG
+  SharedAllocationRecord<void, void>* r = &derived_t::s_root_record;
+
+  char buffer[256];
+
+  SharedAllocationHeader head;
+
+  if (detail) {
+    do {
+      if (r->m_alloc_ptr) {
+        Kokkos::Impl::DeepCopy<HostSpace, MemorySpace>(
+            &head, r->m_alloc_ptr, sizeof(SharedAllocationHeader));
+      } else {
+        head.m_label[0] = 0;
+      }
+
+      // Formatting dependent on sizeof(uintptr_t)
+      const char* format_string;
+
+      if (sizeof(uintptr_t) == sizeof(unsigned long)) {
+        format_string =
+            "%s addr( 0x%.12lx ) list( 0x%.12lx 0x%.12lx ) extent[ 0x%.12lx "
+            "+ %.8ld ] count(%d) dealloc(0x%.12lx) %s\n";
+      } else if (sizeof(uintptr_t) == sizeof(unsigned long long)) {
+        format_string =
+            "%s addr( 0x%.12llx ) list( 0x%.12llx 0x%.12llx ) extent[ "
+            "0x%.12llx + %.8ld ] count(%d) dealloc(0x%.12llx) %s\n";
+      }
+
+      snprintf(buffer, 256, format_string, MemorySpace::execution_space::name(),
+               reinterpret_cast<uintptr_t>(r),
+               reinterpret_cast<uintptr_t>(r->m_prev),
+               reinterpret_cast<uintptr_t>(r->m_next),
+               reinterpret_cast<uintptr_t>(r->m_alloc_ptr), r->m_alloc_size,
+               r->m_count, reinterpret_cast<uintptr_t>(r->m_dealloc),
+               head.m_label);
+      s << buffer;
+      r = r->m_next;
+    } while (r != &derived_t::s_root_record);
+  } else {
+    do {
+      if (r->m_alloc_ptr) {
+        Kokkos::Impl::DeepCopy<HostSpace, MemorySpace>(
+            &head, r->m_alloc_ptr, sizeof(SharedAllocationHeader));
+
+        // Formatting dependent on sizeof(uintptr_t)
+        const char* format_string;
+
+        if (sizeof(uintptr_t) == sizeof(unsigned long)) {
+          format_string = "%s [ 0x%.12lx + %ld ] %s\n";
+        } else if (sizeof(uintptr_t) == sizeof(unsigned long long)) {
+          format_string = "%s [ 0x%.12llx + %ld ] %s\n";
+        }
+
+        snprintf(
+            buffer, 256, format_string, MemorySpace::execution_space::name(),
+            reinterpret_cast<uintptr_t>(r->data()), r->size(), head.m_label);
+      } else {
+        snprintf(buffer, 256, "%s [ 0 + 0 ]\n",
+                 MemorySpace::execution_space::name());
+      }
+      s << buffer;
+      r = r->m_next;
+    } while (r != &derived_t::s_root_record);
+  }
+#else
+  using namespace std::literals::string_literals;
+  Kokkos::Impl::throw_runtime_exception(
+      "SharedAllocationHeader<"s + std::string(MemorySpace::name()) +
+      ">::print_records only works with KOKKOS_ENABLE_DEBUG enabled"s);
+#endif
+}
+
+template <class MemorySpace>
+auto HostInaccessibleSharedAllocationRecordCommon<MemorySpace>::get_record(
+    void* alloc_ptr) -> derived_t* {
```

Comment:
```suggestion
derived_t* HostInaccessibleSharedAllocationRecordCommon<MemorySpace>::get_record(
    void* alloc_ptr) {
```

Reply:
No, there's actually a functional (non-stylistic) reason I did this. Because I have an out-of-line definition of this function template, I can use a nested type alias in the return type specification without qualifying it only if it's a trailing return type (otherwise, C++ wouldn't be parseable because of the `<` template/less-than-operator ambiguity). The suggestion you made won't actually compile on all of the compilers we support.